Sources/MuxPlayerSwift/Fairplay/ContentKeySessionDelegate.swift

```diff
@@ -0,0 +1,153 @@
+        guard let keyURLStr = request.identifier as? String,
+              let keyURL = URL(string: keyURLStr),
+              let assetIDData = keyURLStr.data(using: .utf8)
+        else {
+            print("request identifier was not a key url, this is exceptional for hls")
+            return
+        }
+        
+        let drmToken = lookUpDRMToken(byKeyURL: keyURLStr)
+        guard let drmToken = drmToken else {
+            print("DRM Tokens must be registered when the AVPlayerItem is created, using FairplaySessionManager")
+            return
+        }
+        
+        // get app cert
+        var applicationCertificate: Data?
+        //  the drmtoday example does this by joining a dispatch group, but is this best?
+        let group = DispatchGroup()
+        group.enter()
+        PlayerSDK.shared.fairplaySessionManager.requestCertificate(
+            playbackID: "", // todo - get from sdk caller
+            drmToken: "", // todo - get from sdk caller
+            completion: { result in
+                if let cert = try? result.get() {
+                    applicationCertificate = cert
+                }
+                group.leave()
+            }
+        )
+        group.wait()
+        guard let applicationCertificate = applicationCertificate else {
+            print("failed to get application certificate")
+            return
+        }
+        
+        // step: exchange app cert for SPC using KeyRequest w/completion handler (request wants to know if failed)
+        request.makeStreamingContentKeyRequestData(forApp: applicationCertificate,
+                                                   contentIdentifier: assetIDData,
+                                                   options: [AVContentKeyRequestProtocolVersionsKey: [1]]) { [weak self] spcData, error in
+            guard let self = self else {
+                // todo - log or something?
+                return
+            }
+            
+            guard let spcData = spcData else {
+                print("No SPC Data in spc response")
+                // `error` will be non-nil by contract
+                request.processContentKeyResponseError(error!)
+                return
+            }
+            // step: exchange SPC for CKC using KeyRequest w/completion handler (request wants to know if failed)
+            // todo - drmToken from Asset
+            handleContentKeyResponse(spcData: spcData, drmToken: "", request: request)
+        }
+    }
+    
+    private func handleContentKeyResponse(spcData: Data, drmToken: String, request: AVContentKeyRequest)  {
+        // Send SPC to Key Server and obtain CKC
+        let asset: AVURLAsset // todo - obtain from sdk caller
+        let playbackID: String = "" // todo - obtain from sdk caller / url of asset
+        
+        // todo - DRM Today example does this by joining a DispatchGroup. Is this really preferable??
+        var ckcData: Data? = nil
+        let group = DispatchGroup()
```

Comment:
The `DispatchGroup` here is intended to help make sure these are executed serially, right? 

In that case dispatching to a dedicated serial `DispatchQueue` will avoid the additional bookkeeping required when using `DispatchGroup` with the `enter()`, `leave()`, etc calls.

Reply:
(And the `DispatchGroup` can be dropped entirely then.)

Reply:
Took a closer look and the DRM Today example uses `DispatchGroup` to force asynchronous calls to block and act as synchronous. Probably to help with ordering, I think.

Using a shared serial dispatch queue across the two places (instead of a dedicated serial queue in each place mentioned upthread) where a `DispatchGroup` is used now would give the same ordering enforcement. (And queueing up any other dependent work after they are dispatched would also work.)

Reply:
This sounds like it might be a larger change.

Would you mind if we did this as part of a subsequent PR? This PR is against `project/drm` so nothing will go to GA when we merge this.